### ISSUE TITLE
Handling permissions in ACL-based environments like DSM7

### ISSUE THREAD

**Issue by ludovicchabant** [sabnzbd/sabnzbd#1989]
Hi,

I installed SABnzbd on a Synology running DSM7 and ran into some permission issues. These issues are detailed in [this SynoCommunity ticket][1], with [detailed steps in this message][2].

Based on what [BenjV says][3], DSM7 is using ACLs and, as such, SABnzbd setting permissions using umasks might not work very well. 

The problem may be fixed if SABnzbd was able to recognize that ACLs are in use, and just skip running `set_permissions` altogether in that case. Apparently, it would make the completed downloads naturally inherit their parent directory's permissions.

Worst case, if ACLs can't be auto-detected, maybe the Synology package can install SABnzbd with a default `permissions` config setting set to a special value (like, say, `__acl__` or whatever), which can be checked to ignore setting old-style permissions.

[1]: https://github.com/SynoCommunity/spksrc/issues/4940
[2]: https://github.com/SynoCommunity/spksrc/issues/4940#issuecomment-982919167
[3]: https://github.com/SynoCommunity/spksrc/issues/4940#issuecomment-984004910

**Comment by Safihre** [comment]
@jcfp @sanderjo Help, Linux permissions.. Why again do we set permissions even though no `umask` is set? Aren't permissions fine by default if SABnzbd creates the folders?
Never understood that part, I just know it kinda works.

**Comment by sanderjo** [comment]
> Why again do we set permissions even though no `umask` is set? 

Oh, do we? Where/how?

> Aren't permissions fine by default if SABnzbd creates the folders?

Yes, I would say so.

But the synology DSM7 problems might be different: it seems each process on Synology-DSM7 is now a different user, causing that other process (*arr) can't do everything on stuff delivered by SAB ...

**Comment by puzzledsab** [comment]
Digging back in the commit history indicates that it was initially added to make sure Sab could read the files it had written. It seems to me like the default should be to not set permissions, and then have a switch to turn on the current functionality. Previously it could be set with a command line parameter so maybe it has to be on by default for Sab to start in some cases.

**Comment by thezoggy** [comment]
The preferred solution for dsm7 is to use docker

**Comment by Safihre** [comment]
> The preferred solution for dsm7 is to use docker

.... How does that change anything? SAB breaks the permission system of Synology because it sets low level unix permissions, whether in Docker or not.

**Comment by ludovicchabant** [comment]
> Oh, do we? Where/how?

AFAICT this [happens here][1].

> If ACLs solves that, and Synology people can tell what to set exactly, I suppose SAB can do that.

My understanding is that doing nothing (_i.e._ not setting any permissions) would work because ACLs are inherited by default.



[1]: https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/filesystem.py#L656

**Comment by thezoggy** [comment]
@Safihre I've only seen sab moves result in files inherit parent folder permissions. Is the user explicitly setting it to be something? Or just badly running sab as root or something?

**Comment by Safihre** [comment]
> @Safihre I've only seen sab moves result in files inherit parent folder permissions. Is the user explicitly setting it to be something? Or just badly running sab as root or something? 

No, the problem is correctly identified by @ludovicchabant that SABnzbd forces permissions even if you leave the option for permissions empty. That's quite unexpected from a user point of view.

**Comment by jcfp** [comment]
IIRC there were a bunch of reasons for dealing with permissions, including ensuring sab itself has access, and preventing mere access to the web interface from essentially providing shell access to the system.

On the other hand, setting permissions can obviously interfere with posix access control lists as the file permission bits map to certain acl attributes. Maybe there should be a check for the presence of a default acl on the relevant dir(s) and if so skip setting any permissions other than those explicitly configured by the user?

Relevant dirs being those where sab writes stuff into as part of regular operations, i.e. complete/incomplete/administration. Not sure if the standard python os/stat stuff can check that or if that would need the likes of pylibacl.

**Comment by jcfp** [comment]
Looks like this indeed needs pylibacl. The following works to see if a directory has a default access control list:
```
import posix1e
test_acl = posix1e.ACL(filedef = "/some/dir")
test_acl.valid()
```
Obviously we don't care if the acl is valid but that just seems to be the most straightforward way to check if one is set.

In terms of changes to `filesystem.set_permission` I think we can get away with checking the given path in case it's a directory (and only performing the acl check on the directory 'root' if recursive is active), and looking up the acl on the basedir if it's a file.

**Comment by sanderjo** [comment]
@jcfp Do you understand what the problem and solution is for Synology-DSM7 users?

My guess:
* sab and *arr are diffferent users on Synology-DSM7
* ... and they are not in the same group
* so they can not (see and) delete eachother files 
* solution: SAB should not set dir/file permissions settings if not specified by user in SAB interface ?
* ... plus SAB should do what you describe above with ACL in case ACL is set ... ?
* Does Synology-DSM7 take care of setting up ACL correctly between SAB and *arr?

Is this something we can reproduce on Ubuntu? Different users and ACL activated ... ?

**Comment by jcfp** [comment]
@sanderjo as I understand it, they deploy default acls. Whether those are setup correctly for whatever purpose isn't our problem: it's reasonable to assume that for someone with an advanced setup like this, getting things right is their responsibility too and sab should thus refrain from forcing its own permissions.

If you want to test on a Linux system, create some fresh temp directory and have a go at it:
```
$ umask  # only shows the current umask, notice how it gets overruled by the default acl
0077
$ setfacl -m d:o:rx .
$ getfacl .
# file: .
# owner: username
# group: groupname
user::rwx
group::---
other::---
default:user::rwx
default:group::---
default:other::r-x

$ touch testfile
$ mkdir testdir
$ touch testdir/testfile2
$ ls -la test{dir,file}
-rw----r--  1 username groupname    0 Dec  8 12:43 testfile

testdir:
total 8
drwx---r-x+ 2 username groupname 4096 Dec  8 12:43 .
drwx------+ 5 username groupname 4096 Dec  8 12:43 ..
-rw----r--  1 username groupname    0 Dec  8 12:43 testfile2
```

**Comment by Safihre** [comment]
@jcfp @sanderjo So shouldn't it also be solved if there was just a way to stop SABnzbd from always setting permissions even if the config setting is empty? 
I still don't know if that's really needed..

Or some other way to stop SAB setting permissions?

**Comment by thezoggy** [comment]
Option to skip the logic? As we don't try it on windows

**Comment by sanderjo** [comment]
Yes, like you said in https://github.com/sabnzbd/sabnzbd/issues/1989#issuecomment-985267823 ... don't touch if not set ... ?

**Comment by sanderjo** [comment]
I've been playing with ACL on my Ubuntu, and
* I now understand why it can be handy, and handier than the stone-old rwxrwxrwx
* I know how to set it from CLI, including recursively (see below)
* in the python posix1e module ( http://www.chiark.greenend.org.uk/doc/python-pylibacl/html/module.html / https://pylibacl.k1024.org/module.html ) I don't see a recursive option. Maybe in SAB traverse the subdirector and set the ACL to each item?
* I'll give it a shot in SABnzbd

```
sander@haring1:~$ sudo mkdir /shared_stuff/
sander@haring1:~$ sudo setfacl -m u:sander:rwx /shared_stuff/
sander@haring1:~$ sudo setfacl -m u:pipo:rwx /shared_stuff/
sander@haring1:~$ getfacl /shared_stuff/
getfacl: Removing leading '/' from absolute path names
# file: shared_stuff/
# owner: root
# group: root
user::rwx
user:sander:rwx
user:pipo:rwx
group::r-x
mask::rwx
other::r-x

sander@haring1:~$ echo Hallo > /shared_stuff/my_file.txt
sander@haring1:~$ mkdir /shared_stuff/my_dir1
sander@haring1:~$ echo Blablablalbla > /shared_stuff/my_dir1/blabla.txt
sander@haring1:~$ 



sander@haring1:~$ getfacl /shared_stuff/
getfacl: Removing leading '/' from absolute path names
# file: shared_stuff/
# owner: root
# group: root
user::rwx
user:sander:rwx
user:pipo:rwx
group::r-x
mask::rwx
other::r-x

sander@haring1:~$ getfacl /shared_stuff/my_dir1/
getfacl: Removing leading '/' from absolute path names
# file: shared_stuff/my_dir1/
# owner: sander
# group: sander
user::rwx
group::rwx
other::r-x

sander@haring1:~$ getfacl /shared_stuff/my_file.txt
getfacl: Removing leading '/' from absolute path names
# file: shared_stuff/my_file.txt
# owner: sander
# group: sander
user::rw-
group::rw-
other::r--


sander@haring1:~$ getfacl /shared_stuff/ | setfacl -b -n -M - /shared_stuff/my_dir1/
getfacl: Removing leading '/' from absolute path names

sander@haring1:~$ getfacl /shared_stuff/my_dir1/
getfacl: Removing leading '/' from absolute path names
# file: shared_stuff/my_dir1/
# owner: sander
# group: sander
user::rwx
user:sander:rwx
user:pipo:rwx
group::r-x
mask::rwx
other::r-x

sander@haring1:~$ getfacl /shared_stuff/my_dir1/blabla.txt
getfacl: Removing leading '/' from absolute path names
# file: shared_stuff/my_dir1/blabla.txt
# owner: sander
# group: sander
user::rw-
group::rw-
other::r--

sander@haring1:~$ getfacl /shared_stuff/ | setfacl -R -b -n -M - /shared_stuff/my_dir1/ # -R stands for recursively
getfacl: Removing leading '/' from absolute path names

sander@haring1:~$ getfacl /shared_stuff/my_dir1/
getfacl: Removing leading '/' from absolute path names
# file: shared_stuff/my_dir1/
# owner: sander
# group: sander
user::rwx
user:sander:rwx
user:pipo:rwx
group::r-x
mask::rwx
other::r-x

sander@haring1:~$ getfacl /shared_stuff/my_dir1/blabla.txt
getfacl: Removing leading '/' from absolute path names
# file: shared_stuff/my_dir1/blabla.txt
# owner: sander
# group: sander
user::rwx
user:sander:rwx
user:pipo:rwx
group::r-x
mask::rwx
other::r-x

sander@haring1:~$
```

**Comment by sanderjo** [comment]
Oh, and for @Safihre : it seems getfacl  | setfacl also sets the rwxrwxrwx !

```
sander@haring1:~$ chmod 700 subbie/
sander@haring1:~$ ll subbie/
total 8
drwx------  2 sander sander 4096 Dez  8 22:19 ./
drwxr-xr-x 44 sander sander 4096 Dez  8 22:19 ../
sander@haring1:~$ getfacl alles_en_iedereen/ | setfacl -R -b -n -M - subbie/
sander@haring1:~$ ll subbie/
total 8
drwxrwxrwx  2 sander sander 4096 Dez  8 22:19 ./
drwxr-xr-x 44 sander sander 4096 Dez  8 22:19 ../
sander@haring1:~$
```

**Comment by sanderjo** [comment]
My proposal:
Create an ACL prototype solution in SAB. Get feedback from DSM7 user. Later on, improve where needed.

Implementation:
the value of sabnzbd.ini parameter "permissions" determines what to do:
1. empty => do nothing (Or should we use a separate value like "DoNothing" for this?)
2. octal value (like 777 or 700) => current SAB behaviour
3. "acl1" => set acl only on created directory / file, not on subdirectories
4. "acl2" => set acl on directory and subdirectories. (I think this is needed)

With the above, I think a DSM7 user can determine what works on DSM7.


Feedback @Safihre @jcfp @puzzledsab  ... ?

**Comment by jcfp** [comment]
**Default** ACLs are recursive by nature and require root to set the ownership part, which imho means sab shouldn't bother with setting any permissions at all if a default acl is in effect for relevant directories. Simply create the files in there and let the (file)system deal with permissions and ownership as dictated by the default ACL.

That leaves us with two options: either autodetect default ACLs, or provide some (command line) option to skip setting default permissions. The former is automatic from the user perspective (and unlikely to get it wrong) at the cost of adding a dependency, the latter needs to be explicitly set by the user or package maintainer but doesn't depend on pylibacl. The existing permissions setting -if configured- should have precendence over either.

And it would be nice to be sure this is actually what dsm7 is using, i.e. see relevant getfacl output from such a system.

**Comment by sanderjo** [comment]
> **Default** ACLs are recursive by nature and require root to set the ownership part, which imho means sab shouldn't bother with setting any permissions at all if a default acl is in effect for relevant directories. Simply create the files in there and let the (file)system deal with permissions and ownership as dictated by the default ACL.

That would be nice. I followed https://unix.stackexchange.com/a/1315 , but no luck. So if you have hints ...

```
sander@haring1:~$ sudo mkdir /more_sharing
[sudo] password for sander:
sander@haring1:~$ chmod g+s /more_sharing/
chmod: changing permissions of '/more_sharing/': Operation not permitted
sander@haring1:~$ sudo chmod g+s /more_sharing/
sander@haring1:~$ sudo setfacl -d -m g::rwx /more_sharing/
sander@haring1:~$ sudo setfacl -d -m o::rx /more_sharing/
sander@haring1:~$ ll /more_sharing/
total 8
drwxr-sr-x+  2 root root 4096 Dez 10 11:45 ./
drwxr-xr-x  26 root root 4096 Dez 10 11:45 ../
sander@haring1:~$ date > /more_sharing/my_date.txt
-bash: /more_sharing/my_date.txt: Permission denied
sander@haring1:~$
sander@haring1:~$
sander@haring1:~$ getfacl /more_sharing/
getfacl: Removing leading '/' from absolute path names
# file: more_sharing/
# owner: root
# group: root
# flags: -s-
user::rwx
group::r-x
other::r-x
default:user::rwx
default:group::rwx
default:other::r-x


sander@haring1:~$ sudo setfacl -m u:sander:rwx /more_sharing/
sander@haring1:~$ sudo setfacl -m u:pipo:rwx /more_sharing/
sander@haring1:~$
````
Check:
````
sander@haring1:~$ getfacl /more_sharing/
getfacl: Removing leading '/' from absolute path names
# file: more_sharing/
# owner: root
# group: root
# flags: -s-
user::rwx
user:sander:rwx
user:pipo:rwx
group::r-x
mask::rwx
other::r-x
default:user::rwx
default:group::rwx
default:other::r-x
````
OK, sander and pipo have ACL rights rwx.

And now check if user 'sander' creates a subdir, if another user 'pipo' can alter that subdir (spoiler: no):
```

sander@haring1:~$
sander@haring1:~$ date > /more_sharing/my_date.txt
sander@haring1:~$ mkdir /more_sharing/my_subdir_123
sander@haring1:~$ date > /more_sharing/my_subdir_123/my_date.txt
sander@haring1:~$
sander@haring1:~$ getfacl /more_sharing/my_subdir_123/my_date.txt
getfacl: Removing leading '/' from absolute path names
# file: more_sharing/my_subdir_123/my_date.txt
# owner: sander
# group: root
user::rw-
group::rw-
other::r--

sander@haring1:~$ sudo su - pipo
pipo@haring1:~$
pipo@haring1:~$ date > /more_sharing/my_subdir_123/pipo_date.txt
-su: /more_sharing/my_subdir_123/pipo_date.txt: Permission denied
pipo@haring1:~$
pipo@haring1:~$ rm /more_sharing/my_subdir_123/my_date.txt
rm: remove write-protected regular file '/more_sharing/my_subdir_123/my_date.txt'? y
rm: cannot remove '/more_sharing/my_subdir_123/my_date.txt': Permission denied
pipo@haring1:~$
```

**Comment by ludovicchabant** [comment]
As @jcfp says, ACLs are recursive by default as far as I can tell, so SAB doesn't need to do anything. The only thing it might do is auto-detect whether ACLs are active or not, but even that can be avoided because the package installer (SynoCommunity in my case) can set the proper defaults for SAB's configuration.

So unless there are other use-cases to cover, you could get by with just (1) some option to not set any permissions and mayyyybe (2) auto-detect when to do that.  Whether (1) is done by leaving permissions empty (thus changing previous behaviour) or by setting a special value (like, say, `acl` or whatever instead of an octal value) is up to you. Regardless of the choice, I'll probably go back to the SynoCommunity project so that it sets that config setting on install anyway.

**Comment by Safihre** [comment]
@ludovicchabant I make the SynoCommunity package for SABnzbd 😊

**Comment by ludovicchabant** [comment]
@Safihre well that makes my job here even easier 😆

**Comment by sanderjo** [comment]
FWIW (@Safihre ):

a plain `python3 -m pip install pylibacl` does not work ... see https://github.com/iustin/pylibacl/issues/11#issuecomment-994698938 ... you need a `sudo apt install libacl1-dev` beforehand ... othwerwise pip-wheel errors.

Weird? Or normal?

**Comment by sanderjo** [comment]
Some notes on ACL and Ubuntu/Linux:

* getfacl / setfacl are part of the package `acl`, which is not installed by default
* python package pylibacl needs a header file to get installed (see earlier comment)
* dirty trick: "ls -al" shows on a directory and a file a `+` at the end of the rwx if ACL is set. So that is a poor man's method to see if ACL is set if the above tools are not available.

```
sander@brixit:~$ ll /sharing/
total 16
drwxrwxr-x+  2 root   root   4096 dec 15 13:39 ./
drwxr-xr-x  21 root   root   4096 dec 15 13:18 ../
-rw-rwxr--+  1 sander sander   28 dec 15 13:38 date.txt*
-rw-rw-r--   1 sander sander   28 dec 15 13:39 no_acl.txt


```

**Comment by jcfp** [comment]
> a plain `python3 -m pip install pylibacl` does not work ...
> Weird? Or normal?

That's normal for modules that ship an extension that needs some library, see (comments in) sab's own requirement.txt.

**Comment by jcfp** [comment]
WIP

**Comment by Safihre** [comment]
@jcfp I actually bought a new Synology NAS for testing with DSM 7. Already have an ancient one with DSM 6.2.

Did Synology not use the ACL on DSM 6?

**Comment by Safihre** [comment]
@jcfp With my new NAS I think I can confirm what @ludovicchabant has been saying since the start: if Linux-permissions are not forced by the application (e.g. SABnzbd) it will automatically inherit the ACL permissions of the parent.
I logged in with SSH and did a simple:
```
os.mkdir("test")
```
They show inherited permissions (grey = inherited):
![image](https://user-images.githubusercontent.com/5703454/146604035-6f71c52c-a7c4-4be9-bddb-9920b45eb6f3.png)

However, on the SABnzbd created folders (permissions setting in SABnzbd was left empty) the permissions are explicit:
![image (5)](https://user-images.githubusercontent.com/5703454/146604133-96882530-832d-4e11-bea6-1b5b234d6fac.png)

**Comment by Safihre** [comment]
Closed by 0bb543222359c4f7ab124c3fa51f5a6e12bf97f2
